# Supervised Learning on Breast Cancer Dataset using Python

### The objective of this assignment is to evaluate the understanding and application of supervised learning techniques on a real-world dataset.

### 2 = load data

In [7]:
# Import libraries
from sklearn.datasets import load_breast_cancer
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load dataset
data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = pd.Series(data.target)

# Check first 5 rows
X.head()


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

### 3 = Explore

In [8]:
# Check size of data
X.shape, y.shape


((569, 30), (569,))

In [9]:
# See feature names
X.columns

Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error', 'fractal dimension error',
       'worst radius', 'worst texture', 'worst perimeter', 'worst area',
       'worst smoothness', 'worst compactness', 'worst concavity',
       'worst concave points', 'worst symmetry', 'worst fractal dimension'],
      dtype='object')

In [10]:
# Understand target labels
data.target_names

array(['malignant', 'benign'], dtype='<U9')

In [11]:
# Check class balance
y.value_counts()

1    357
0    212
Name: count, dtype: int64

In [12]:
# Statistical summary
X.describe()


mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst radius  \
count     569.000000              569.000000  ...    569.000000   
mean        0.181162                0.062798  ...     16.269190   
std         0.027414                0.007060  ...      4.833242   
min         0.106000                0.049960  ...      7.930000   
25%         0.161900                0.057700  ...     13.010000   
50%         0.179200                0.061540  ...     14.970000   
75%         0.195700                0.066120  ...     18.790000   
max         0.304000                0.097440  ...     36.040000   

       worst texture  worst perimeter   worst area  worst smoothness  \
count     569.000000       569.000000   569.000000        569.000000   
mean       25.677223       107.261213   880.583128          0.132369   
std         6.146258        33.602542   569.356993          0.022832   
min        12.020000        50.410000   185.200000          0.071170   
25%        21.080000        84.110000   515.300000          0.116600   
50%        25.410000        97.660000   686.500000          0.131300   
75%        29.720000       125.400000  1084.000000          0.146000   
max        49.540000       251.200000  4254.000000          0.222600   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.254265         0.272188              0.114606   
std             0.157336         0.208624              0.065732   
min             0.027290         0.000000              0.000000   
25%             0.147200         0.114500              0.064930   
50%             0.211900         0.226700              0.099930   
75%             0.339100         0.382900              0.161400   
max             1.058000         1.252000              0.291000   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.290076                 0.083946  
std          0.061867                 0.018061  
min          0.156500                 0.055040  
25%          0.250400                 0.071460  
50%          0.282200                 0.080040  
75%          0.317900                 0.092080  
max          0.663800                 0.207500  

[8 rows x 30 columns]

### 4 =preprocessing

In [13]:
X.isnull().sum()


mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
dtype: int64

In [14]:
## To make all columns comparable.
## Feature scaling was applied using StandardScaler to normalize feature values, ensuring fair 
## contribution of all features to distance-based and gradient-based classifiers.

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


In [15]:
## The dataset was split into training and testing sets to train the model and evaluate its ability to generalize to new, unseen data.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)


## Train model

### 1 = Logistic Regression Classification

In [16]:
##Logistic Regression was imported to perform binary classification of tumors as malignant or benign. Accuracy score, confusion matrix, and classification report were used to evaluate the performance of the model.

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


In [17]:
# Create model
logreg = LogisticRegression()

In [18]:
# Train model
# The Logistic Regression model was trained using the training dataset, allowing it to learn patterns between tumor features and their corresponding diagnoses.
logreg.fit(X_train, y_train)

LogisticRegression()

In [19]:
# Predict
y_pred_logreg = logreg.predict(X_test)

In [ ]:
# Evaluate prediction
accuracy_logreg = accuracy_score(y_test, y_pred_logreg)

# matrix of evaluation
conf_matrix_logreg = confusion_matrix(y_test, y_pred_logreg)

# detailed performance report
report_logreg = classification_report(y_test, y_pred_logreg)

In [ ]:
# prediction accuracy
print("Logistic Regression Accuracy:", accuracy_logreg)

Logistic Regression Accuracy: 0.9736842105263158


In [25]:
print("Confusion Matrix:\n", conf_matrix_logreg)

Confusion Matrix:
 [[41  2]
 [ 1 70]]


In [ ]:
# Report
print("Classification Report:\n", report_logreg)

Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.95      0.96        43
           1       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114



#Logistic Regression was implemented and evaluated using accuracy, confusion matrix, and classification report, achieving an accuracy of 97.37%.


### 2 = Decision Tree

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Create Decision Tree model
dtree = DecisionTreeClassifier(random_state=42)

# Train the model
dtree.fit(X_train, y_train)

# Predict on test data
y_pred_dtree = dtree.predict(X_test)

# Evaluate the model
accuracy_dtree = accuracy_score(y_test, y_pred_dtree)
conf_matrix_dtree = confusion_matrix(y_test, y_pred_dtree)
report_dtree = classification_report(y_test, y_pred_dtree)

print("Decision Tree Accuracy:", accuracy_dtree)
print("Confusion Matrix:\n", conf_matrix_dtree)
print("Classification Report:\n", report_dtree)


In [26]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [27]:
# Create Decision Tree model
dtree = DecisionTreeClassifier(random_state=42)

In [28]:
# Train the model
dtree.fit(X_train, y_train)


DecisionTreeClassifier(random_state=42)

In [29]:
# Predict on test data
y_pred_dtree = dtree.predict(X_test)

In [30]:
# Evaluate the model
accuracy_dtree = accuracy_score(y_test, y_pred_dtree)
conf_matrix_dtree = confusion_matrix(y_test, y_pred_dtree)
report_dtree = classification_report(y_test, y_pred_dtree)

In [31]:
print("Decision Tree Accuracy:", accuracy_dtree)

Decision Tree Accuracy: 0.9473684210526315


In [32]:
print("Confusion Matrix:\n", conf_matrix_dtree)

Confusion Matrix:
 [[40  3]
 [ 3 68]]


In [33]:
print("Classification Report:\n", report_dtree)

Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.96      0.96      0.96        71

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114



In [34]:
## Decision Tree Classifier predicted cancer with an accuracy of 93%, and the confusion matrix and classification report show detailed performance metrics for each class.

### 3 = Random Forest Classifier

In [35]:
# Import Random Forest and evaluation tools
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [36]:
# Create Random Forest model
rf = RandomForestClassifier(random_state=42)

In [37]:
# Train the model
rf.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

In [38]:
# Predict on test data
y_pred_rf = rf.predict(X_test)

In [39]:
# Evaluate the model
accuracy_rf = accuracy_score(y_test, y_pred_rf)
conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)
report_rf = classification_report(y_test, y_pred_rf)

In [40]:
print("Random Forest Accuracy:", accuracy_rf)

Random Forest Accuracy: 0.9649122807017544


In [41]:
print("Confusion Matrix:\n", conf_matrix_rf)

Confusion Matrix:
 [[40  3]
 [ 1 70]]


In [42]:
print("Classification Report:\n", report_rf)

Classification Report:
               precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114



In [43]:
#The Random Forest model performs very well, correctly identifying most cancer cases while rarely misclassifying benign cases, achieving an overall accuracy of 96%.

### 4 = Support Vector Machine (SVM)

# Create SVM model
svm_model = SVC(kernel='linear', random_state=42)

# Train the model
svm_model.fit(X_train, y_train)

# Predict on test data
y_pred_svm = svm_model.predict(X_test)

# Evaluate the model
accuracy_svm = accuracy_score(y_test, y_pred_svm)
conf_matrix_svm = confusion_matrix(y_test, y_pred_svm)
report_svm = classification_report(y_test, y_pred_svm)

# Print results
print("SVM Accuracy:", accuracy_svm)
print("Confusion Matrix:\n", conf_matrix_svm)
print("Classification Report:\n", report_svm)

In [44]:
# Import SVM and evaluation tools
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [45]:
# Create SVM model
svm_model = SVC(kernel='linear', random_state=42)

In [46]:
# Train the model
svm_model.fit(X_train, y_train)

SVC(kernel='linear', random_state=42)

In [47]:
# Predict on test data
y_pred_svm = svm_model.predict(X_test)

In [48]:
# Evaluate the model
accuracy_svm = accuracy_score(y_test, y_pred_svm)
conf_matrix_svm = confusion_matrix(y_test, y_pred_svm)
report_svm = classification_report(y_test, y_pred_svm)

In [49]:
print("SVM Accuracy:", accuracy_svm)

SVM Accuracy: 0.956140350877193


In [50]:
print("Confusion Matrix:\n", conf_matrix_svm)

Confusion Matrix:
 [[41  2]
 [ 3 68]]


In [53]:
print("Classification Report:\n", report_svm)

Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.95      0.94        43
           1       0.97      0.96      0.96        71

    accuracy                           0.96       114
   macro avg       0.95      0.96      0.95       114
weighted avg       0.96      0.96      0.96       114





SVM classified breast tumors with 96% accuracy, effectively distinguishing malignant and benign cases.**


In [55]:
## SVM classified breast tumors with 96% accuracy, effectively distinguishing malignant and benign cases.


### 5 = k-Nearest Neighbors (k-NN) Classifier

In [56]:
# Import k-NN and evaluation tools
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [57]:
# Create k-NN model
knn = KNeighborsClassifier(n_neighbors=5)

In [58]:
# Train the model
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [59]:
# Predict on test data
y_pred_knn = knn.predict(X_test)


In [60]:
# Evaluate the model
accuracy_knn = accuracy_score(y_test, y_pred_knn)
conf_matrix_knn = confusion_matrix(y_test, y_pred_knn)
report_knn = classification_report(y_test, y_pred_knn)

In [61]:
print("k-NN Accuracy:", accuracy_knn)

k-NN Accuracy: 0.9473684210526315


In [62]:
print("Confusion Matrix:\n", conf_matrix_knn)

Confusion Matrix:
 [[40  3]
 [ 3 68]]


In [63]:
print("Classification Report:\n", report_knn)

Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.93      0.93        43
           1       0.96      0.96      0.96        71

    accuracy                           0.95       114
   macro avg       0.94      0.94      0.94       114
weighted avg       0.95      0.95      0.95       114



In [64]:
## The k-Nearest Neighbors (k-NN) classifier was used to classify breast tumors.
## The model achieved an accuracy of 95%, showing reliable performance in distinguishing malignant and benign cases.

In [ ]:
# Algorithm             	Accuracy
# Logistic Regression	    97%
# Decision Tree	            93%
# Random Forest	            96%
# Support Vector Machine	96%
# k-Nearest Neighbors	    95%

## Model Comparison

The performance of five supervised learning algorithms was compared using accuracy as the evaluation metric. Logistic Regression achieved the highest accuracy of 97%, followed by Random Forest and Support Vector Machine with 96% each. k-Nearest Neighbors achieved 95%, while Decision Tree showed comparatively lower performance at 93%.

## Conclusion



In this assessment, supervised learning techniques were applied to the breast cancer dataset to classify tumors as malignant or benign. Among the models tested, Logistic Regression performed the best and was selected as the final model due to its high accuracy and consistent performance.